File: JAVA/18_Stack/BOJ_04_4949_균형잡힌세상.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        while (true) {
            int tst = 0;
            String str = br.readLine();
            if (str.equals(".")) break;
            Stack<Character> stack = new Stack<>();
            for (int i = 0; i < str.length(); i++) {
                if (str.charAt(i) == '(' || str.charAt(i) == '[') {
                    stack.push(str.charAt(i));
                } else if (str.charAt(i) == ')') {
                    if (stack.empty() || stack.peek() != '(') {
                        sb.append("no\n");
                        tst++;
                        break;
                    } else stack.pop();
                }
                else if (str.charAt(i) == ']') {
                    if (stack.empty() || stack.peek() != '[') {
                        sb.append("no\n");
                        tst++;
                        break;
                    } else stack.pop();
                }
            }
            if (tst == 0) {
                if (stack.empty()) sb.append("yes\n");
                else sb.append("no\n");
            }
        }
        System.out.println(sb);
    }
}

File: C/18_Stack/BOJ_04_4949_균형잡힌세상.c
#include <stdio.h>
#include <string.h>
int top = -1;
void push(char a, char* stack) {
	stack[++top] = a;
}
void pop(char* stack) {
	stack[top--] = 0;
}
main() {
	while (1) {
		char str[101], stack[101];
		int tst = 0;
		top = -1;
		gets(str); //scanf("%[^\n]s", str)를 사용해봤는데 오류가 난다
		if (str[0] == '.' && strlen(str) == 1) break;
		for (int j = 0; j < strlen(str); j++) {
			if (str[j] == '(' || str[j] == '[') push(str[j], stack);
			else if (str[j] == ')') {
				if (top == -1 || stack[top] == '[') {
					tst = 1;
					break;
				}
				else if (stack[top] == '(') pop(stack);
			}
			else if (str[j] == ']') {
				if (top == -1 || stack[top] == '(') {
					tst = 1;
					break;
				}
				else if (stack[top] == '[') pop(stack);
			}
		}
		if (tst == 0 && top == -1) printf("yes\n");
		else printf("no\n");
	}
}